### ISSUE TITLE
object/callback lifetimes?

### ISSUE THREAD

**Issue by kitlith** [jnr/jnr-ffi#231]
to be clear, this is more asking a question because i can't find anything on it more than reporting an actual issue.

When you a pass a java callback to a native library, how long is a reference taken? I'm dealing with a ffi layer that has asynchronous callbacks, when would jnr-ffi/jffi stop holding a reference to them?

what's come up in my head are the following:
 - holds a reference for the duration of the function call, then releases it afterwards
   - this would break down for the async callbacks, but i guess i could work around it by attaching a reference to the callback to the returned future or something to prevent the garbage collector from going after it.
 - ~~holds a reference until the callback is called, then drops it~~
   - this wouldn't work with the sort example, unless it was combined with the former rule. meh.
 - holds a reference forever, effectively leaking the memory used for the callback.
   - that's... not ideal, and probably not what you're doing, but it would work in all cases.

m'sorry if i'm overthinking this.

**Comment by basshelal** [comment]
From my somewhat little experience using JNR this is how I've seen it happen:

Consider this basic example:

```java
public class MyObject {

    private MyCallback callback; // you need a strong reference to the callback! 

    public setCallback(MyCallback callback) {
        this.callback = callback;
        MyLibrary.setCallback(callback); // calls native code and sets the callback there 
    }

    @Delegate
    public interface MyCallback {
        public void invoke(Pointer data);
    }

}
```
The callback should be kept in native code forever, so long as you don't do `setCallback(null)` which would be how you would remove a callback.

Here's the trick though, if an instance of `MyObject` had set a callback and then that instance was later GC'd at some point because it was out of scope, **the callback will still be called**, leading to a memory leak.

Thus, you need to have a *"destructor"* that will remove this callback before it's GC'd. 

You basically have to manage/enforce this yourself since implementing `java.lang.Object.finalize()` is a bad idea (the docs there explain why). 

You can *probably* try some fancy way of managing this using caches or some of the stuff in the `java.lang.ref` package but shit gets complex and unpredictable very quickly in doing so, so try to keep it simple, I speak from a very bad experience regarding a similar issue (non JNR).

I don't know if this is intentional behavior of JNR (it probably is), but this is my experience with callbacks, I was concerned at first like you so I wrote a bunch of tests to see what would happen and this is what I found.

Hope this helps and hopefully I'm not speaking complete nonsense as I haven't read the internals of how this is handled, @headius  would know more about that

**Comment by kitlith** [comment]
Yeah, okay, that sounds basically exactly like how i would end up dealing with the behaviour in my first hypothetical. Except... would have to ensure that futures are completed before letting them get GC'd. blargh.

**Comment by basshelal** [comment]
Yeah my bad experience was with a similar thing on Android.

The best solution for this problem is some kind of global collection (like a cache) that keeps strong references to the futures which also keep strong references to the objects that keep a strong reference to the callback. Once the future is done you call your destructor which will remove the callback and then, remove the future from the cache, thus ready for GC and no memory leaks (but do test this thoroughly).

How you determine a future is complete will differ by implementation, RxJava provides this kind of functionality out of the box with *"onComplete"* callbacks, but RxJava is quite complicated and is a little too bloated in scope and functionality.

I have previously successfully avoided RxJava and used timed polling to do this by having a single thread check every couple hundred millis to check if a future is complete and remove them from the collection, the `java.util.Timer` class has what you need.

See [this](https://stackoverflow.com/a/20135721/7175336), I eventually solved my problem quite well this way, but be careful with growing complexity, *that's* where RxJava will actually help.

Hope this helps 😊

**Comment by kitlith** [comment]
What I ended up doing instead was just storing the callbacks themselves in a `ConcurrentHashMap.newKeySet()` before calling the ffi method, and then having the callback remove itself from the set as the very last thing it does. Future ended up being just a normal CompletableFuture, with no special handling.

**Comment by kitlith** [comment]
something like c#'s GCHandle could be really useful here (and for other objects as well), but i'm not sure if that's something that jnr can just implement on top of java.

**Comment by basshelal** [comment]
If you absolutely must have more control then you could use `java.lang.Object.finalize()` but be aware of its potential issues especially across different VM implementations.

Honestly, I think this is a smaller issue than you're making it out to be unless I'm missing something major. Let me know if that is the case, maybe provide some code examples.

**Comment by kitlith** [comment]
it's probably the anxiety :tm:, though finalize has nothing to do with what I was referring to with GCHandle.

**Comment by kitlith** [comment]
In case anyone comes looking at this issue in the future:

When i mentioned GCHandle, what I was looking for was ObjectReferenceManager: https://github.com/jnr/jnr-ffi/blob/a6271ada96075823b6e4417698cdae8d041e8d6b/src/main/java/jnr/ffi/ObjectReferenceManager.java
